[Monitoring] Introducing Logs UI

chrisronline opened the pull request:
Resolves https://github.com/elastic/stack-monitoring/issues/27

## Summary
This is a short to mid term logging implementation within Stack Monitoring. The long term solution to embed the [Logs UI components](https://www.elastic.co/guide/en/infrastructure/guide/current/logs-ui-overview.html) within Stack Monitoring directly; however, these are some blockers on this front (Links coming soon...)

This solution involves showing, at most, 10 logs in a table with a link off to the Logs UI if the user wants to dive deeper. The idea behind this is to try and show the user recent, relevant logs that might help them debug their problem easier. If that doesn't help them enough, give them a link to the Logs UI that will allow them to view all the logs.

We also added something on the overview cluster page that shows a list of the _types_ of logs (server, deprecation, slowlog, etc) and a count of the logs for each _level_ (debug, log, warn, etc) of that _type_

## Blockers
Currently, we have a couple hard blockers on this approach:

1) ~~https://github.com/elastic/kibana/issues/23316 - We need the ability to "deep link" into the Logs UI so we can only show the logs the user really cares about~~ Resolved by #31653

2) ~~https://github.com/elastic/kibana/issues/31657 - The Logs UI needs to the ability to read from the Monitoring ES cluster (instead of the default one configured through Kibana)~~

3) #30792 - **We decided to merge in spite of this work not done yet, see [this comment](https://github.com/elastic/kibana/pull/31275#issuecomment-482503619)** The Logs UI needs to support CCS by default. This doesn't affect the code in this PR, but clicking on the Logs UI link won't work properly until this is resolved.

## Screenshots
<img width="1833" alt="screen shot 2019-02-15 at 10 54 26 am" src="https://user-images.githubusercontent.com/56682/52868156-8e691580-3110-11e9-85e7-4f2ebe41c9d1.png">
<img width="481" alt="screen shot 2019-02-15 at 10 54 31 am" src="https://user-images.githubusercontent.com/56682/52868157-8e691580-3110-11e9-9e8c-a9e4d3d700aa.png">
<img width="1835" alt="screen shot 2019-02-15 at 10 56 05 am" src="https://user-images.githubusercontent.com/56682/52868158-8e691580-3110-11e9-914a-2c5c534fb530.png">
<img width="473" alt="fb_no_es" src="https://user-images.githubusercontent.com/56682/55106723-02022900-50a6-11e9-93a5-1defa930576b.png">
<img width="1856" alt="no_logs_for_cluster" src="https://user-images.githubusercontent.com/56682/55106725-02022900-50a6-11e9-89c4-882795d99d7d.png">
<img width="1809" alt="no_logs_for_index" src="https://user-images.githubusercontent.com/56682/55106726-02022900-50a6-11e9-99e3-c5279d9d6d4a.png">
<img width="1860" alt="no_logs_for_node" src="https://user-images.githubusercontent.com/56682/55106727-029abf80-50a6-11e9-997b-8b5d02d93165.png">
<img width="464" alt="no_fb" src="https://user-images.githubusercontent.com/56682/55106728-029abf80-50a6-11e9-8816-43d6415b8273.png">
<img width="473" alt="none_in_time_period" src="https://user-images.githubusercontent.com/56682/55106729-029abf80-50a6-11e9-9ce9-b807b6b0bc8d.png">



## TODO
- [x] ~~Localization~~ Added
- [x] ~~Tests~~ Added unit and api integration.
- [x] ~~Figure out if `size: 10` should be a constant or configurable?~~ This defaults to `10`, but is configured through `xpack.monitoring.elasticsearch.logFetchCount` to a max of `50`
- [x] ~~Make the `filebeat-*` indices a constant (or maybe from config too?)~~ This is a constant, but not configurable by users.

## Testing

### Default
1) Install filebeat and configure the [elasticsearch module](https://www.elastic.co/guide/en/beats/filebeat/master/filebeat-module-elasticsearch.html). We need filebeat to collect the elasticsearch logs and send them to the _monitoring_ cluster (which is probably the same cluster as the production one for most developer testing setups but worth noting if you have more than one cluster running). I'd enable, at least, the _server_ and _deprecation_ log.
2) Run Kibana and [configure Monitoring](https://www.elastic.co/guide/en/elasticsearch/reference/current/configuring-monitoring.html) using either internal collection or Metricbeat (preferably Metricbeat)
3) Go to Stack Monitoring in Kibana.

### CCS
1) Start a separate, second ES instance connected to the same cluster as the first one. Ensure it runs on a separate http and tcp port.
2) Change filebeat configuration to point to the second ES instance
3) Configure CCS from the first ES instance to the second instance. Use the tcp port for the configuration.
4) Test again. The logs should show up in the UI

chrisronline committed:
Initial implementation

chrisronline committed:
More logs UI work

elasticmachine commented:
Pinging @elastic/stack-monitoring

ycombinator commented:
Very cool to see this starting to come together!

For the `size`, my vote would be to make it a configurable (as opposed to a constant), in case users want to preview more or less. We could default it to `10` and perhaps only allow a max of `50` or something?

For the `filebeat-*` index pattern, my vote would be to start off making this a constant. Let's see if there's demand from users to make it configurable and then, if so, convert it to a configurable. From an API perspective, it'll be easier to go from a constant -> setting than the other way around.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/898/)

chrisronline committed:
Remove unnecessary code

chrisronline committed:
Add support to build a logs url based on the cluster and/or node uuid

chrisronline committed:
Deep link directly

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1653/)

chrisronline committed:
Update link

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2029/)

chrisronline committed:
Use CCS to access remote filebeat data

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5490/)

chrisronline committed:
Fix existing tests

chrisronline committed:
Add log specific api integration tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5622/)

chrisronline committed:
Localization

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5671/)

chrisronline committed:
Adding more localization

chrisronline committed:
Adding unit tests for logs ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5815/)

chrisronline committed:
Client side unit tests, configuration for log fetch count, adding visual callout for why we can't detect logs

chrisronline committed:
Remove debug

chrisronline committed:
Fix localization issue

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5890/)

chrisronline committed:
Update tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5936/)

spalger commented on x-pack/test/api_integration/apis/monitoring/logs/cluster.js:
```diff
@@ -0,0 +1,39 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import expect from 'expect.js';
```
Need to merge master and rewrite this import before merging: https://github.com/elastic/kibana/pull/33761

ycombinator commented:
Good stuff! Some feedback on functionality, all pretty minor:

1. > Showing the most recent logs, up to 10 total logs

   For a bit of extra clarity, I think on the Cluster Overview page we should say "Showing the most recent logs **for this cluster,** up to 10 total logs" and on the Node Overview page we should say, "Showing the most recent logs **for this node**, up to 10 total logs".

3. Most, if not all, log entries will contain a `node.name` field. At least on the Cluster Overview page's Log Preview table it might be useful to pull this out into its own column. That way if there are interwoven logs from multiple nodes being shown in the same table it'll be easier to distinguish which ones came from which node.

4. Happy to defer this to a follow up PR but it might useful to provide some filter (not text search) elements at the top of the table, e.g. to allow the user to only see `ERROR` level logs or logs from a certain component, etc.

5. In the "No filebeat indices found." case, we point the users to the Filebeat setup documentation, which is great. I wonder if we should also say that they should configure their Filebeat Elastiscearch output to their monitoring cluster?

6. There's an edge case where if you narrow the time window in the time picker to some window in the past when there were no logs, we show the "No filebeat indices found" message and direct the user to setup filebeat. I wonder if we should say something like "No filebeat indices found for the selected time window" and direct the user to either change the time window or setup filebeat. I recall seeing similar messaging about not seeing data in the selected time window in another part of the Monitoring UI (but can't recall exactly where) so we might want to be consistent with the language if we choose to change it here.

7. Some log entries have an `index.name` field. This would allow us to show the Logs Preview table on the Index Overview page. WDYT about adding it there?

8. You probably planned to do this already so more of a reminder/+1: Please have @gchaps check through all the language in the UI introduced in this PR.

9. The Log UI seems to be using slightly different styling, e.g. the text font, vertical alignment of text within each row, and the row hover bgcolor. Should we make the preview table use the same styling so it gives the user a bit of a sense of continuity when they navigate over to the Logs UI to see more?

chrisronline committed:
PR feedback

chrisronline commented:
Thanks for the great feedback @ycombinator!

Comments inline:

> For a bit of extra clarity, I think on the Cluster Overview page we should say "Showing the most recent logs for this cluster, up to 10 total logs" and on the Node Overview page we should say, "Showing the most recent logs for this node, up to 10 total logs". Or you could shorten it a bit to say "Showing the most recent 10 log entries for this cluster/node".

Done

> Most, if not all, log entries will contain a node.name field. At least on the Cluster Overview page's Log Preview table it might be useful to pull this out into its own column. That way if there are interwoven logs from multiple nodes being shown in the same table it'll be easier to distinguish which ones came from which node.

Yup, good point. I made this change so the cluster log table shows the name name too.

> Happy to defer this to a follow up PR but it might useful to provide some filter (not text search) elements at the top of the table, e.g. to allow the user to only see ERROR level logs or logs from a certain component, etc.

Let's defer that for now. I don't want to add too much in this PR.

> In the "No filebeat indices found." case, we point the users to the Filebeat setup documentation, which is great. I wonder if we should also say that they should configure their Filebeat Elastiscearch output to their monitoring cluster?

Agreed. Updated the text.

> There's an edge case where if you narrow the time window in the time picker to some window in the past when there were no logs, we show the "No filebeat indices found" message and direct the user to setup filebeat. I wonder if we should say something like "No filebeat indices found for the selected time window" and direct the user to either change the time window or setup filebeat. I recall seeing similar messaging about not seeing data in the selected time window in another part of the Monitoring UI (but can't recall exactly where) so we might want to be consistent with the language if we choose to change it here.

Nice find. I added an edge case where we not only search for `filebeat-*` indices in the specified time range, but we search without regard to time. If we find _some_ documents outside the time period, we show a relevant message now.

> Some log entries have an index.name field. This would allow us to show the Logs Preview table on the Index Overview page. WDYT about adding it there?

Good idea. I added this.

> You probably planned to do this already so more of a reminder/+1: Please have @gchaps check through all the language in the UI introduced in this PR.

Yes thanks for the reminder.

> The Log UI seems to be using slightly different styling, e.g. the text font, vertical alignment of text within each row, and the row hover bgcolor. Should we make the preview table use the same styling so it gives the user a bit of a sense of continuity when they navigate over to the Logs UI to see more?

I don't disagree here, but I'd rather defer this for a later PR, but happy to discuss if someone feels strongly about it.

chrisronline commented:
Bad rebase. Sorry for the ping folks

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6852/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6857/)

cachedout commented:
I like it!

I have a few suggestions, all of which are just thoughts about simple improvements in follow-up PRs and not necessarily appropriate for this PR. Happy to copy them to an issue if that's preferable:

* I think it would go a long way if we could add some basic colorization to logs. For example, simply being able to display anything logged at a `WARN` level in red would really help people see critical issues quickly.

* How are timestamps being handled across timezones? Are we just displaying whatever Filebeat is telling us about the timestamp?

* Oddly, timestamps don't line up for me exactly. For example, here is the entry in the monitoring application: `March 26, 2019 8:59:32 PM | INFO | Server | updating index lifecycle policy [filebeat-8.0.0]` and here is the log entry in my console: `[2019-03-26T14:59:30,319][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [Mikes-MBP] updating index lifecycle policy [filebeat-8.0.0]`.

* It's not clear to me what the `Type` column indicates. What values other than `Server` are expected?

* Have we thought at all about whether or not to display a message to users who haven't configured filebeat that by doing so they can view logs? This might help drive adoption.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6865/)

chrisronline commented:
Thanks for the great feedback @cachedout! I'll address them inline here:

> I think it would go a long way if we could add some basic colorization to logs. For example, simply being able to display anything logged at a WARN level in red would really help people see critical issues quickly.

Hmm I agree, but I wonder if the logging UI provides anything like this, so we can be consistent.

> How are timestamps being handled across timezones? Are we just displaying whatever Filebeat is telling us about the timestamp? Oddly, timestamps don't line up for me exactly. For example, here is the entry in the monitoring application: March 26, 2019 8:59:32 PM | INFO | Server | updating index lifecycle policy [filebeat-8.0.0] and here is the log entry in my console: [2019-03-26T14:59:30,319][INFO ][o.e.x.i.a.TransportPutLifecycleAction] [Mikes-MBP] updating index lifecycle policy [filebeat-8.0.0].

I'm not entirely sure here honestly. If you go into the logging UI, do you see the same inconsistency?

> It's not clear to me what the Type column indicates. What values other than Server are expected?

ES emits various types of logs, as [documented here for filebeat](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-module-elasticsearch.html#configuring-elasticsearch-module) and we're just showing those various types

> Have we thought at all about whether or not to display a message to users who haven't configured filebeat that by doing so they can view logs? This might help drive adoption.

Yup, we have this now, even though I didn't include it in the PR description (it came in after it I wrote it), but we have various states. I'll drop those in, but you should see them them as you go through the setup process.

chrisronline committed:
Update import

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6909/)

ycombinator commented:
>> The Log UI seems to be using slightly different styling, e.g. the text font, vertical alignment of text within each row, and the row hover bgcolor. Should we make the preview table use the same styling so it gives the user a bit of a sense of continuity when they navigate over to the Logs UI to see more?

> I don't disagree here, but I'd rather defer this for a later PR, but happy to discuss if someone feels strongly about it.

I don't think we should merge this feature into `master` unless we have consistency in styling. Otherwise we risk shipping something with an inconsistent UX for users.

Do you want to defer this work to a later PR to reduce the scope of this one? If so, we could merge these PRs into a feature branch first, then merge the feature branch to `master`.

ycombinator commented:
>> Some log entries have an index.name field. This would allow us to show the Logs Preview table on the Index Overview page. WDYT about adding it there?

> Good idea. I added this.

This doesn't seem to be working for me. When I ran the following aggs query, it showed me 2 indices in `filebeat-*`:

```
POST filebeat-8.0.0-*/_search
{
  "size": 0, 
  "aggs": {
    "indices": {
      "terms": {
        "field": "elasticsearch.index.name",
        "size": 10
      }
    }
  }
}
```

However, if I visit one of those indices in the Stack Monitoring UI, I don't see any logs. I checked that the time period was large enough.

ycombinator commented:
I'm able to see logs on the Index page. Something was probably screwed up in my setup.

chrisronline committed:
Format the count to avoid a huge string of numbers

ycombinator reviewed (approved):
Functionally this PR LGTM.

I'd suggest consulting the Kibana design folks, and @gchaps in particular for verbiage, and getting a code review from another UI developer.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7022/)

cachedout commented:
> I'm not entirely sure here honestly. If you go into the logging UI, do you see the same inconsistency?

The mystery deepens. For the same log entry, here is the timestamp in Stack Monitoring:
`March 26, 2019 9:50:32 PM`

Here it is in the log viewer:
`Mar 26, 2019 @ 21:50:30.988`

Even if it's rounding up, it's still off by 1s which is very odd. Also, we should note the inconsistency between time stamp formats here. (Personally, I prefer the log viewer format but that is personal preference.)

chrisronline commented:
Nice find @cachedout!

We're using the wrong timestamp field. Will fix it up in the next push

chrisronline committed:
Use @timestamp instead

chrisronline committed:
Handle scenario where the time period is not right but the type exists

chrisronline committed:
Update jest tests

justinkambic commented:
It's worth mentioning that @chrisronline and I did find a case where the UI behaved in a non-ideal way; if Filebeat is configured to consume incorrect files (I was pointing my FB at ES's `logs` directory and giving it a pattern of `*.log` when it should've had `{SPECIFIC_LOG_FILE_NAME}.json`), the UI was showing the logs box for the cluster, but no text.

I'd be fine if the UI were adapted to either show nothing or show a message suggesting what might be happening, but I admittedly don't have enough domain knowledge to say whether those solutions are ideal. If action is needed I'd say doing it in a follow-up PR is fine.

Aside from that I've concluded that the functionality works well for me. I'm in the process of performing code review, so far I haven't found anything worth changing but I'll likely review it more than once as it's a sizable diff.

chrisronline committed:
Update api tests

chrisronline commented:
@justinkambic I'm actually not able to reproduce that locally. I see an appropriate error message in that state. Can you consistently reproduce that?

justinkambic commented:
> I'm actually not able to reproduce that locally. I see an appropriate error message in that state. Can you consistently reproduce that?

Let me see if I still can.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7075/)

justinkambic reviewed (approved):
LGTM

chrisronline commented:
Pinging @gchaps and @snide.

@gchaps Can you provide any feedback on the copy additions for this PR?

@snide We have one unresolved design concern. In a nutshell, we're not sure if we should attempt to make the stack monitorings logs ui look and feel similar to the infra logs ui, or should we try and make a clear distinction? This UI will lack significant functionality so I'm leading towards ensuring it looks intentionally different, but happy to zoom about it or discuss more here.

yaronp68 commented:
@chrisronline any plans to breakdown logs by node, so as I user when I look at node table or node view I see the number of logs and can get the list of logs etc

chrisronline commented:
@yaronp68 Do you mean have like the logs by type view from the overview page on the node page itself, so you'd see like `68 debug logs`, `7 deprecation logs`?

cachedout commented:
Hi @yaronp68 Did you have an answer for @chrisronline on his question above?

cachedout commented:
Looks like good progress is happening over on the logs viewer side. Here's the latest: https://github.com/elastic/kibana/pull/34455

snide commented:
Hey, i missed this original ask I think when I was on vacation.

> We have one unresolved design concern. In a nutshell, we're not sure if we should attempt to make the stack monitorings logs ui look and feel similar to the infra logs ui, or should we try and make a clear distinction? This UI will lack significant functionality so I'm leading towards ensuring it looks intentionally different, but happy to zoom about it or discuss more here.

I think in a perfect world you'd just USE the logs UI and link directly over to it. The same way that APM or Infra link over to it. And if the functionality doesn't match what you need, the ui there should be adjusted to be more flexible.

Know that's a pretty loaded answer and likely you're way too far down a path already, but from a product perspective I think it makes sense to view logs, whatever they may be in one set of tooling so the user always has the same experience. Might be worth pinging @weltenwort

chrisronline commented:
Thanks for weighing in @snide!

We're doing that, but we also want to provide some inline logs to help the user debug problems they are seeing in the monitoring graphs. It's meant to provide a quick glimpse while letting the user go the logs ui for a deeper dive. I think we want to provide this glimpse until we can fully integrate the logs viewer into the stack monitoring app itself

chrisronline committed:
Text changes

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9293/)

chrisronline commented:
@elastic/stack-monitoring WDYT about merging this before waiting for the outstanding blocker (https://github.com/elastic/kibana/issues/30792) simply because this only affect a CCS-related setup, which we aren't even recommending?

chrisronline committed:
Add periods

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9332/)

chrisronline committed:
Update tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9341/)

cachedout commented:
> WDYT about merging this before waiting for the outstanding blocker (#30792) simply because this only affect a CCS-related setup, which we aren't even recommending?

👍 from me.

yaronp68 commented:
@chrisronline Great job! Would it be possible to add a filter above the table to ease finding the log file one is looking for?
Another thing is filtering by log type (if possible)

cachedout commented:
@yaronp68 I think we're hoping to get this in and then take feature requests in follow-up PRs. If you'd like to file an issue with those requests, we can get them triaged and prioritized. Thanks!

chrisronline commented:
@yaronp68 Let's talk about follow ups here: https://github.com/elastic/stack-monitoring/issues/27

gchaps commented:
@chrisronline Text changes look good.

chrisronline commented:
@ycombinator @justinkambic  I know you already LGTM the PR, but did you want to re-review with some recent text changes? Or do you still feel good about it?

ycombinator commented:
If the only changes after my last review were text changes, the PR still LGTM.

chrisronline commented:
Backport:
7.x: https://github.com/elastic/kibana/commit/e87682c5f8268e5841c880fc1cf412abd9145052

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9657/)
